File: js/binarySearch.js
let idx = 0

const compare = (arr, target) => {

  if(arr.length === 0) {
    if(arr[0] === target) {
      return idx
    } else {
      return -1
    }
  }

  const mid = Math.floor(arr.length / 2)
  const pivot = arr[mid]
  const right = arr.slice(mid + 1)
  const left = arr.slice(0, mid)
  if (pivot === target) {
    idx += mid
    return idx
  } else if (pivot > target) {
    return compare(left, target)
  } else {
    idx += mid + 1
    return compare(right, target)
  }
}

File: rs/binary_search.rs
fn main() {
  print!("{}", binary_search(&[1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 5));
}
fn binary_search(arr: &[i32], target: i32) -> i32 {
  let mut left = 0;
  let mut right = arr.len() - 1;
  while left <= right {
    let mid = left + (right - left) / 2;
    if arr[mid] == target {
      return mid as i32;
    } else if arr[mid] < target {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }
  -1
}